File: src/main/java/at/yawk/wm/dock/module/DockBuilder.java
package at.yawk.wm.dock.module;

import at.yawk.wm.PeriodBuilder;
import at.yawk.wm.Scheduler;
import at.yawk.wm.dock.Dock;
import at.yawk.wm.hl.Monitor;
import at.yawk.wm.style.FontManager;
import at.yawk.wm.ui.*;
import at.yawk.wm.x.GlobalResourceRegistry;
import at.yawk.wm.x.Screen;
import at.yawk.wm.x.Window;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import javax.inject.Inject;
import javax.inject.Singleton;
import lombok.SneakyThrows;
import org.slf4j.Logger;

/**
 * @author yawkat
 */
@Singleton
public class DockBuilder implements RenderElf {
    private static final Logger log = org.slf4j.LoggerFactory.getLogger(DockBuilder.class);
    @Inject Monitor monitor;
    @Inject DockConfig config;
    @Inject Screen screen;
    @Inject GlobalResourceRegistry globalResourceRegistry;
    @Inject Scheduler scheduler;
    @Inject FontManager fontManager;
    @Inject PeriodBuilder periodBuilder;

    private Dock dock;

    void start(DockBootstrap bootstrap) {
        log.info("Initializing dock...");

        dock = new Dock(screen, config.getBackground());
        globalResourceRegistry.register(dock);
        dock.setBounds(monitor.getX(), monitor.getY(), monitor.getWidth(), config.getHeight());

        setupWidgets(bootstrap);

        bootstrap.getDockStartListeners().forEach(Runnable::run);

        dock.show();
        log.info("Dock initialized");
    }

    private void decorate(Widget widget) {
        if (widget instanceof TextWidget) {
            ((TextWidget) widget).setTextHeight(config.getHeight());
        } else if (widget instanceof IconWidget) {
            ((IconWidget) widget).setTargetHeight(config.getHeight());
        } else if (widget instanceof FlowCompositeWidget) {
            ((FlowCompositeWidget) widget).getWidgets().forEach(this::decorate);
        }
    }

    void addLeft(Widget widget) {
        widget.setOrigin(Origin.TOP_LEFT);
        decorate(widget);
        dock.getLeft().addWidget(widget);
    }

    void addRight(Widget widget) {
        widget.setOrigin(Origin.TOP_RIGHT);
        decorate(widget);
        dock.getRight().addWidget(widget);
    }

    @Override
    public void render() {
        if (dock != null) {
            dock.render();
        }
    }

    //////// SETUP ////////

    @SuppressWarnings("unchecked")
    @SneakyThrows // lazy
    private void setupWidgets(DockBootstrap bootstrap) {
        List<Widget> widgets = bootstrap.getWidgets();
        log.info("Visiting {} widgets...", widgets.size());
        Collections.sort(widgets, Comparator.comparingInt(
                w -> w.getClass().getAnnotation(DockWidget.class).priority()));
        for (Widget widget : widgets) {
            Class<? extends Widget> widgetClass = widget.getClass();
            DockWidget annotation = widgetClass.getAnnotation(DockWidget.class);

            DockWidget.Position position = annotation.position();
            if (position == DockWidget.Position.LEFT) {
                addLeft(widget);
            } else {
                addRight(widget);
            }

            widget.init();

            periodBuilder.scan(widget);
        }

        periodBuilder.flush();
    }

    public Window getWindow() {
        return dock.getWindow();
    }
}

File: src/main/java/at/yawk/wm/dock/module/DockBuilder.kt
package at.yawk.wm.dock.module

import at.yawk.wm.di.PerMonitor
import at.yawk.wm.PeriodBuilder
import at.yawk.wm.dock.Dock
import at.yawk.wm.hl.Monitor
import at.yawk.wm.ui.FlowCompositeWidget
import at.yawk.wm.ui.IconWidget
import at.yawk.wm.ui.Origin
import at.yawk.wm.ui.RenderElf
import at.yawk.wm.ui.TextWidget
import at.yawk.wm.ui.Widget
import at.yawk.wm.x.GlobalResourceRegistry
import at.yawk.wm.x.Window
import at.yawk.wm.x.XcbConnector
import org.slf4j.LoggerFactory
import java.util.Comparator
import java.util.function.Consumer
import javax.inject.Inject

/**
 * @author yawkat
 */
@PerMonitor
class DockBuilder @Inject constructor(
    private val monitor: Monitor,
    private val xcbConnector: XcbConnector,
    private val globalResourceRegistry: GlobalResourceRegistry
) : RenderElf {
    private lateinit var dock: Dock

    private lateinit var periodBuilder: PeriodBuilder

    fun start(bootstrap: DockBootstrap) {
        log.info("Initializing dock...")
        this.periodBuilder = bootstrap.periodBuilder
        dock = Dock(xcbConnector.screen, DockConfig.background.awt)
        globalResourceRegistry.register(dock)
        dock.setBounds(monitor.x, monitor.y, monitor.width, DockConfig.height)
        setupWidgets(bootstrap)
        bootstrap.dockStartListeners.forEach(Consumer { obj: Runnable -> obj.run() })
        dock.show()
        log.info("Dock initialized")
    }

    private fun decorate(widget: Widget) {
        when (widget) {
            is TextWidget -> widget.textHeight = DockConfig.height
            is IconWidget -> widget.targetHeight = DockConfig.height
            is FlowCompositeWidget ->
                widget.widgets.forEach { decorate(it) }
        }
    }

    fun addLeft(widget: Widget) {
        widget.origin = Origin.TOP_LEFT
        decorate(widget)
        dock.left.addWidget(widget)
    }

    fun addRight(widget: Widget) {
        widget.origin = Origin.TOP_RIGHT
        decorate(widget)
        dock.right.addWidget(widget)
    }

    override fun render() {
        if (::dock.isInitialized) {
            dock.render()
        }
    }

    //////// SETUP ////////
    private fun setupWidgets(bootstrap: DockBootstrap) {
        var widgets = bootstrap.widgets
        log.info("Visiting {} widgets...", widgets.size)
        widgets = widgets.sortedWith(Comparator.comparingInt { it.javaClass.getAnnotation(DockWidget::class.java).priority })
        for (widget in widgets) {
            val widgetClass: Class<out Widget> = widget.javaClass
            val annotation = widgetClass.getAnnotation(DockWidget::class.java)
            val position: DockWidget.Position = annotation.position
            if (position == DockWidget.Position.LEFT) {
                addLeft(widget)
            } else {
                addRight(widget)
            }
            widget.init()
        }
        periodBuilder.flush()
    }

    val window: Window
        get() = dock.window

    companion object {
        private val log = LoggerFactory.getLogger(DockBuilder::class.java)
    }
}